Change debug output to be a button.

rogerl50 opened the pull request:
@Ingegneus Resolves #1372 

The open and close buttons seem sluggish at times to me, which makes no sense since all they do is toggle a classlist element.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1374/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1374-pr-test`.

96LawDawg commented:
This changes other behavior, perhaps unintentionally?  On production, when you hover over to display, all the clickable arrows in a multi-step routine are collapsed.  In this PR, they are expanded.  I prefer the collapsed view for editing as it is easier to see things and then to just drill down to what I need and not expand every function by default.

A couple of very minor things:  The X to close button shifts by a couple of pixels when you move the scroll bar.  Not sure why that is, but it is slightly annoying.  Also, I don't know that VTT blue is the right color for the button.  There are often other actual room buttons in that same vicinity that are VTT blue.  That might be slightly confusing (it was to me at first). I think she Show Debug button and the close button should be some other color.  I don't have a good suggestion.  Maybe the inverse (white with VTT blue text)?

I didn't notice any sluggishness, even when showing very complicated routines, but I didn't play with it very much.

Ingegneus commented:
I too noticed the "sluggishness" when opening the debug output (something like 0.3s). I don't have a problem with the vtt blue, especially because I think adding a bit of clearance around the button, would sufficiently separate it from the actual game. also maybe use the close_circle_white icon from the vtt font? it looks a bit nicer. other than that it does exactly what I want 👍

robartsd commented:
> The open and close buttons seem sluggish at times to me, which makes no sense since all they do is toggle a classlist element.

You've made other changes to the stylesheet making it more expensive for the browser to initially render the debug output. I doubt there would have been performance changes if you simply replaced `#jeLogWrapper:hover` with `#jeLogWrapper.active` and provided buttons to toggle the class. Opening in collapsed view would almost certainly load faster. Styling `#jeLog` with `display:inline` instead of `display:block` may also be a factor.

ArnoldSmith86 commented:
Am I the only one who would prefer a normal toolbar button in the top right to toggle it?

rogerl50 commented:
> You've made other changes to the stylesheet making it more expensive for the browser to initially render the debug output. I doubt there would have been performance changes if you simply replaced `#jeLogWrapper:hover` with `#jeLogWrapper.active` and provided buttons to toggle the class. Opening in collapsed view would almost certainly load faster. Styling `#jeLog` with `display:inline` instead of `display:block` may also be a factor.

Thanks. I think this was almost all due to the fact that the whole tree opened initially. Once I fixed that, things seem fine. I also changed the display mode for `#jeLog`.

rogerl50 commented:
> This changes other behavior, perhaps unintentionally? On production, when you hover over to display, all the clickable arrows in a multi-step routine are collapsed. In this PR, they are expanded. I prefer the collapsed view for editing as it is easier to see things and then to just drill down to what I need and not expand every function by default.

Doh. Fixed.

> 
> A couple of very minor things: The X to close button shifts by a couple of pixels when you move the scroll bar. Not sure why that is, but it is slightly annoying. 

I have no idea why. Perhaps @robartsd can comment.

>Also, I don't know that VTT blue is the right color for the button. There are often other actual room buttons in that same vicinity that are VTT blue. That might be slightly confusing (it was to me at first). I think she Show Debug button and the close button should be some other color. I don't have a good suggestion. Maybe the inverse (white with VTT blue text)?

I changed the button colors, and also the text of the close button per another comment.

rogerl50 commented:
> > This changes other behavior, perhaps unintentionally? On production, when you hover over to display, all the clickable arrows in a multi-step routine are collapsed. In this PR, they are expanded. I prefer the collapsed view for editing as it is easier to see things and then to just drill down to what I need and not expand every function by default.
> 
> Doh. Fixed.
> 
> > A couple of very minor things: The X to close button shifts by a couple of pixels when you move the scroll bar. Not sure why that is, but it is slightly annoying.
> 
> I have no idea why. Perhaps @robartsd can comment. Also, if there were a way to position that sticky button on top of the upper right corner of the `#jeLog` element, that would be ideal.
> 
> > Also, I don't know that VTT blue is the right color for the button. There are often other actual room buttons in that same vicinity that are VTT blue. That might be slightly confusing (it was to me at first). I think she Show Debug button and the close button should be some other color. I don't have a good suggestion. Maybe the inverse (white with VTT blue text)?
> 
> I changed the button colors, and also the text of the close button per another comment.

rogerl50 commented:
> Am I the only one who would prefer a normal toolbar button in the top right to toggle it?

Probably not, but you're the only one who thought of it! Is there general agreement that this is a better approach?

96LawDawg commented:
> > Am I the only one who would prefer a normal toolbar button in the top right to toggle it?
> 
> Probably not, but you're the only one who thought of it! Is there general agreement that this is a better approach?

Yes.  That.  Do that.

96LawDawg reviewed (approved):
The bug icon works also.  Good work on this PR and a nice improvement overall.  Approved through test.

robartsd commented:
A bit nitpicky, but the on/off styling for the highlighter and debug should probably match.

On the highlight toggle, the slash goes from top left to bottom right and does not go all the way to the corners. It also indicates the current state rather than the state pushing the button will trigger. The debug toggle has a slash from top right to bottom left (corner to corner) and indicates the state that pushing the button will trigger.

rogerl50 commented:
> A bit nitpicky, but the on/off styling for the highlighter and debug should probably match.
> 
> On the highlight toggle, the slash goes from top left to bottom right and does not go all the way to the corners. It also indicates the current state rather than the state pushing the button will trigger. The debug toggle has a slash from top right to bottom left (corner to corner) and indicates the state that pushing the button will trigger.
> 
That's a reasonable point. I can easily reverse the state and the orientation of the line. If you look at the code, you'll see that I used some css magic to create the diagonal line (since there is no material icon with a slash through it for "no bug"). Any ideas for how I can do that and not have the line corner to corner?

robartsd commented:
> Any ideas for how I can do that and not have the line corner to corner?

A line could be generated in CSS by positioning a `::before` or `::after` pseudo element such that a border of the element would be the desired line.

ArnoldSmith86 commented:
Maybe we change all toggle buttons to retain their icon but toggle between normal and dark blue background?

rogerl50 commented:
> Maybe we change all toggle buttons to retain their icon but toggle between normal and dark blue background?

This would apply to the Zoom, Edit mode, Wide, Highlight, and Debug buttons (and also potentially to the F-Key order button). In all cases, we could make the background some other color when the button is in its non-default state.

This does seem easier and more consistent than the current situation.

96LawDawg commented:
> > Maybe we change all toggle buttons to retain their icon but toggle between normal and dark blue background?
> 
> This would apply to the Zoom, Edit mode, Wide, Highlight, and Debug buttons (and also potentially to the F-Key order button). In all cases, we could make the background some other color when the button is in its non-default state.
> 
> This does seem easier and more consistent than the current situation.

Just so long as there is enough visual difference between the dark blue and the grey of unavailable icons.  Right now, as I hover over an icon and it switches to dark blue, it is pretty close to the gray of the unavailable icons.  A slightly different shade of blue might make those easier to distinguish.  But I definitely support the concept.

ArnoldSmith86 commented:
Inverting might be an option but I'm not sure if I like it:

![image](https://user-images.githubusercontent.com/73538315/198297481-741967d3-fe07-4094-b60f-97cd7730eb76.png)

```CSS
text-shadow: 0 1px 2px #ccc;
color: var(--VTTblue);
background: white;
```

rogerl50 commented:
> ```css
> text-shadow: 0 1px 2px #ccc;
> color: var(--VTTblue);
> background: white;
> ```

I thought about that as well. I think it just looks too stark/contrasty. Another possibility is VTTblueLight:
<img width="166" alt="Screen Shot 2022-10-27 at 9 56 56 AM" src="https://user-images.githubusercontent.com/63379656/198305025-30c72131-9566-426e-b40a-a2bf35c5a64f.png">
but I'm not sure I like that either.
```css
text-shadow: 0 1px 2px #ccc;
background: var(--VTTblueLight);
```

96LawDawg commented:
I’m on mobile so can’t check this out, but how about a new color that is halfway between VTT Blue and VTT Dark Blue. According to a color blender site I’ve used that would be #164682. Or if you want it a little darker #133E76 which would be like 2/3 of the way between them.

rogerl50 commented:
> I’m on mobile so can’t check this out, but how about a new color that is halfway between VTT Blue and VTT Dark Blue. According to a color blender site I’ve used that would be #164682. Or if you want it a little darker #133E76 which would be like 2/3 of the way between them.

#164682 looks like this:
<img width="158" alt="Screen Shot 2022-10-27 at 10 30 43 AM" src="https://user-images.githubusercontent.com/63379656/198315143-f6ec6202-d9f0-4a01-b124-ed688b854620.png">
Not that much better than dark blue IMO.

rogerl50 commented:
Another idea is to change the color of the icon itself. We could get rid of the yellow color on Duplicate and use that for the non-default state for the other buttons. We know that's visible.

robartsd commented:
I like yellow for on. Widget highlight currently defaults to on, the other toggles default to off.

rogerl50 commented:
Here's what that looks like...

96LawDawg commented:
I like the yellow.  That is a very good solution.  Approved (again).

rogerl50 commented:
> Widget highlight default doesn’t matter to me either way, but I think on should be the default. The others defaulting to off is also correct as far as my vote goes.

On *is* the default. I guess the flashlight should not be yellow, then. But of course that means it will be yellow when the (yellow) highlighting is off. Awkward.

mousewax commented:
My two cents: 
1¢: Flashlight on = highlight on. 
1¢: On = yellow, off = white (like it is now) but the difference needs to be greater.
1¢: ![Artboard 1](https://user-images.githubusercontent.com/12822213/198390825-8386c1bf-bb6a-4814-a13c-b5f95eed9f44.png)

robartsd commented on client/css/jsonedit.css:
```diff
@@ -426,6 +426,12 @@ body.jsonEdit.jeZoomOut #topSurface {
 }
 
+/* For the zoom, edit, wide, highlight, and debug "on" states */
+[id=jeTopButtons] button.onState {
```
Is `[id=jeTopButtons]` being used instead of `#jeTopButtons` to reduce selector specificity here?

rogerl50 replied:
I was mirroring the syntax slightly further up in that file. Is there a difference between `[id=jeTopButtons]` and `#jeTopButtons`?

robartsd replied:
> I was mirroring the syntax slightly further up in that file. Is there a difference between `[id=jeTopButtons]` and `#jeTopButtons`?

CSS rules are applied according to [specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity). More specific rules trump less specific rules regardless of where the rules appear. Although `#jeTopButtons` and `[id=jeTopButtons]` select the same element, `#jeTopButton` is considered more specific because `[id=jeTopButtons]` is evaluated as an attribute selector rather than the special id selector (same weight as a class selector).

rogerl50 replied:
The only use of `#jeTopButtons` as a selector appears directly above the code you commented on. I don't see why replacing all occurrences of `[id=jeTopButtons]` with `#jeTopButtons` would cause any issues, but all of that was there before I started working on this, and I've been afraid to change it given my level of understanding of CSS.

robartsd replied:
I suppose using attribute selectors to select an ancestor of the target widget by id is simply best practice because `#jeTopButtons button.onState` would higher calculated specificity (1-1-1) than `#je_toggleZoom.onState` (1-1-0) even though `#je_toggleZoom.onState` is actually more specific.

rogerl50 commented:
I believe this is ready for code review.

ArnoldSmith86 commented on client/js/jsonedit.js:
```diff
@@ -1788,6 +1807,12 @@ function jeLoggingRoutineEnd(variables, collections) {
   --jeLoggingDepth;
   if(!jeLoggingDepth) {
     $('#jeLog').innerHTML = jeLoggingHTML + '</div></div>';
+    // Enable the "Show Routine Debug Output" button the first time a routine is executed.
```
Not a big deal but the button is longer called that.

rogerl50 replied:
Fixed, thanks.

ArnoldSmith86 commented on client/js/jsonedit.js:
```diff
@@ -2072,8 +2097,8 @@ function jeShowCommands() {
     if(command.context == undefined) {
       const name = (typeof command.name == 'function' ? command.name() : command.name);
       const icon = (typeof command.icon == 'function' ? command.icon() : command.icon);
-      const isMaterial = String(icon).match(/^[^[]/) ? 'material' : '';
-      commandText += `<button class='top ${isMaterial}' id='${command.id}' title='${name}' ${!command.show || command.show() ? '' : 'disabled'}>${icon}</button>`;
+      const classes = typeof command.classes == 'function' ? command.classes() : command.classes || '';
+      commandText += `<button class='top ${classes}' id='${command.id}' title='${name}' ${!command.show || command.show() ? '' : 'disabled'}>${icon}</button>`;
```
Hmm, I think I would prefer to still detect `material` automatically and then append the result of `command.classes()`. What do you think?

ArnoldSmith86 reviewed (approved):
I have no problem merging this as is. I just added a few comments. We can think about icon changes after this PR...

rogerl50 commented on client/css/jsonedit.css:
```diff
@@ -435,6 +426,24 @@ body.jsonEdit.jeZoomOut #topSurface {
+#je_toggleHighlight.onState {
+  position: relative;
+}
+#je_toggleHighlight.onState::before {
+  position: absolute;
+  height: 22px;
+  width: 17px;
+  content: '';
+  top: 1px;
+  left: 4px;
+  border: 2px solid yellow;
+  border-radius: 20%;
+}
```
Based on your comment in Discord, I think I'll leave this. Happy to consider other suggestions any time, though.
